yt_dlp/networking/impersonate.py

```diff
@@ -0,0 +1,119 @@
+from __future__ import annotations
+
+from abc import ABC
+from typing import Any, Optional, Tuple
+
+from .common import RequestHandler, register_preference
+from .exceptions import UnsupportedRequest
+from ..compat.types import NoneType
+from ..utils.networking import std_headers
+
+ImpersonateTarget = Tuple[Optional[str], Optional[str], Optional[str], Optional[str]]
+
+
+def _target_within(target1: ImpersonateTarget, target2: ImpersonateTarget):
+    for i in range(0, min(len(target1), len(target2))):
+        if (
+            target1[i]
+            and target2[i]
+            and target1[i] != target2[i]
+        ):
+            return False
+
+    return True
+
+
+class ImpersonateRequestHandler(RequestHandler, ABC):
+    """
+    Base class for request handlers that support browser impersonation.
+
+    This provides a method for checking the validity of the impersonate extension,
+    which can be used in _check_extensions.
+
+    Impersonate targets are defined as a tuple of (client, version, os, os_vers).
+    Note: Impersonate targets are not required to define all fields (except client).
+
+    The following may be defined:
+     - `_SUPPORTED_IMPERSONATE_TARGET_TUPLES`: a tuple of supported targets to impersonate.
+        Any Request with an impersonate target not in this list will raise an UnsupportedRequest.
+        Set to None to disable this check.
+     - `_SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP`: a dict mapping supported targets to custom targets.
+        This works similar to `_SUPPORTED_IMPERSONATE_TARGET_TUPLES`.
+
+    Note: Only one of `_SUPPORTED_IMPERSONATE_TARGET_TUPLE_MAP` and `_SUPPORTED_IMPERSONATE_TARGET_TUPLES` can be defined.
```

Comment:
Only the dict version is needed, no?

Reply:
my original intention was to leave this open since you may not need the mapping in some situations

Reply:
We can always add new API in future. So try to keep it to just what we know we need.